### PR TITLE
Add rate limiter, rename maxConn into inFlightReq

### PR DESCRIPTION
**Pull request by mpl**
### What does this PR do?

The rate limiting middleware had been removed at the beta1 version of Traefik v2, because its behaviour seemed unintuitive at times, and made us doubt its usefulness. It is now being reintroduced based on a simpler and clearer package (golang.org/x/time/rate), which results in a more satisfying behaviour and throughput. It also allowed us to change the configuration for it.

In addition, as the maxConn middleware did not exactly reflect what it was about, it is being renamed as inflightReq. Indeed, it is not about the number of currently opened connections (since there can be many requests using the same keep-alive connection), but rather about the number of requests being simultaneously processed.

### Motivation

To have a simpler and more intuitive rate limiting middleware.

### More

- [x] Added/updated tests
- [x] Added/updated documentation

##### Rate limiting

```toml
# BEFORE
[http.middlewares]
  [http.middlewares.example.rateLimit]
    extractorFunc = "client.ip"
    
    [http.middlewares.example.rateLimit.rateSet.rate0]
      period = "10s"
      average = 100
      burst = 200
    
    [http.middlewares.example.rateLimit.rateSet.rate1]
      period = "3s"
      average = 5
      burst = 10

# AFTER
[http.middlewares]
  [http.middlewares.example.rateLimit]
    average = 100
    burst = 50
  [http.middlewares.example.rateLimit.sourceCriterion]
    requestheadername = "foobar"
    requesthost = true
    [http.middlewares.example.rateLimit.sourceCriterion.ipStrategy]
      depth =  3
      excludedIPs = ["foo", "bar"]
```

```yaml
# BEFORE
- "traefik.http.middlewares.example.ratelimit.extractorfunc=client.ip"
- "traefik.http.middlewares.example.ratelimit.rateset.rate0.period=10s"
- "traefik.http.middlewares.example.ratelimit.rateset.rate0.average=100"
- "traefik.http.middlewares.example.ratelimit.rateset.rate0.burst=200"
- "traefik.http.middlewares.example.ratelimit.rateset.rate1.period=3s"
- "traefik.http.middlewares.example.ratelimit.rateset.rate1.average=5"
- "traefik.http.middlewares.example.ratelimit.rateset.rate1.burst=10"

# AFTER
- "traefik.http.middlewares.example.ratelimit.average=100"
- "traefik.http.middlewares.example.ratelimit.burst=50"
- "traefik.http.middlewares.example.ratelimit.sourcecriterion.requestheadername=foobar"
- "traefik.http.middlewares.example.ratelimit.sourcecriterion.requesthost=true"
- "traefik.http.middlewares.example.ratelimit.sourcecriterion.ipstrategy.depth=3"
- "traefik.http.middlewares.example.ratelimit.sourcecriterion.ipstrategy.excludedips=foo,bar"
```

##### Inflight Requests

```toml
# BEFORE
[http.middlewares]
  [http.middlewares.example.maxConn]
    amount = 42
    extractorFunc = "foobar"

[http.middlewares]
  [http.middlewares.example.inFilghtReq]
    amount = 100
  [http.middlewares.example.inFilghtReq.sourceCriterion]
    requestHost = true
    requestHeaderName = "foobar"
    [http.middlewares.example.inFilghtReq.sourceCriterion.ipStrategy]
      depth = 3
      excludedIPs = ["foo", "bar"]
```

```yaml
# BEFORE
- "traefik.http.middlewares.example.maxconn.amount=100"
- "traefik.http.middlewares.example.maxconn.extractorfunc=request.host"

# AFTER
- "traefik.http.middlewares.example.inflightreq.amount=100"
- "traefik.http.middlewares.example.inflightreq.sourcecriterion.ipstrategy.depth=3"
- "traefik.http.middlewares.example.inflightreq.sourcecriterion.ipstrategy.excludedips=foo,bar"
- "traefik.http.middlewares.example.inflightreq.sourcecriterion.requestheadername=foobar"
- "traefik.http.middlewares.example.inflightreq.sourcecriterion.requesthost=true"

```

Co-authored-by: Julien Salleyron <julien.salleyron@gmail.com>
Co-authored-by: Jean-Baptiste Doumenjou <jb.doumenjou@gmail.com>

### CODE REVIEW

File: pkg/config/dynamic/middlewares.go [hunk 1]
```diff
@@ -259,15 +258,17 @@ type IPWhiteList struct {
...
+type InFlightReq struct {
+	Amount          int64            `json:"amount,omitempty" toml:"amount,omitempty" yaml:"amount,omitempty"`
+	SourceCriterion *SourceCriterion `json:"sourceCriterion,omitempty" toml:"sourceCriterion,omitempty" yaml:"sourceCriterion,omitempty"`
 }
 
-// SetDefaults Default values for a MaxConn.
-func (m *MaxConn) SetDefaults() {
-	m.ExtractorFunc = "request.host"
+// SetDefaults Default values for a InFlightReq.
+func (m *InFlightReq) SetDefaults() {
```
Comment (mmatur) [inline pkg/config/dynamic/middlewares.go]: `m` should be renamed

File: pkg/middlewares/inflightreq/inflight_req.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+	handler, err := connlimit.New(next, sourceMatcher, config.Amount)
+	if err != nil {
+		return nil, fmt.Errorf("error creating connection limit: %v", err)
+	}
+
+	return &inFlightReq{handler: handler, name: name}, nil
+}
+
+func (mc *inFlightReq) GetTracingInformation() (string, ext.SpanKindEnum) {
```
Comment (mmatur) [inline pkg/middlewares/inflightreq/inflight_req.go]: `mc` should be renamed

File: pkg/config/dynamic/middlewares.go (hunk 1 above)
Comment (mpl) [reply pkg/config/dynamic/middlewares.go]: done

File: pkg/middlewares/inflightreq/inflight_req.go (hunk 2 above)
Comment (mpl) [reply pkg/middlewares/inflightreq/inflight_req.go]: done

**Pull request merged**